fontforgeexe/cvpalettes.c

```diff
@@ -819,100 +817,281 @@ static void CVPolyStar(CharView *cv) {
-			            &GIcon_smallcorner, &GIcon_smalltangent,
-			            &GIcon_smallscale, &GIcon_smallrotate,
-			            &GIcon_smallflip, &GIcon_smallskew,
-			            &GIcon_small3drotate, &GIcon_smallperspective,
-			            &GIcon_smallrect, &GIcon_smallpoly,
-			            &GIcon_smallelipse, &GIcon_smallstar };
-    static GImage *spirosmalls[] = { &GIcon_smallpointer, &GIcon_smallmag,
-				    &GIcon_smallpencil, &GIcon_smallhand,
-			            &GIcon_smallknife, &GIcon_smallruler,
-			            &GIcon_smallspiroright, NULL,
-				    &GIcon_smallspirocurve, &GIcon_smallspirog2curve,
-			            &GIcon_smallspirocorner, &GIcon_smallspiroleft,
-			            &GIcon_smallscale, &GIcon_smallrotate,
-			            &GIcon_smallflip, &GIcon_smallskew,
-			            &GIcon_small3drotate, &GIcon_smallperspective,
-			            &GIcon_smallrect, &GIcon_smallpoly,
-			            &GIcon_smallelipse, &GIcon_smallstar };
     static const unichar_t _Mouse[][9] = {
 	    { 'M', 's', 'e', '1',  '\0' },
 	    { '^', 'M', 's', 'e', '1',  '\0' },
 	    { 'M', 's', 'e', '2',  '\0' },
 	    { '^', 'M', 's', 'e', '2',  '\0' }};
-    int i,j,norm, mi;
+    int i,j,sel,norm, mi;
     int tool = cv->cntrldown?cv->cb1_tool:cv->b1_tool;
     int dither = GDrawSetDither(NULL,false);
     GRect temp;
     int canspiro = hasspiro(), inspiro = canspiro && cv->b.sc->inspiro;
-    GImage *(*buttons)[2] = inspiro ? spirobuttons : normbuttons;
+    GImage* (*buttons)[14][2] = (inspiro ? spirobuttons : normbuttons);
     GImage **smalls = inspiro ? spirosmalls : normsmalls;
 
-    normbuttons[3][1] = canspiro ? &GIcon_spiroup : &GIcon_spirodisabled;
+    normbuttons[0][3][1] = canspiro ? &GIcon_spiroup : &GIcon_spirodisabled;
 
     GDrawPushClip(pixmap,r,&old);
     GDrawFillRect(pixmap,r,GDrawGetDefaultBackground(NULL));
     GDrawSetLineWidth(pixmap,0);
-    for ( i=0; i<sizeof(normbuttons)/sizeof(normbuttons[0])-1; ++i ) for ( j=0; j<2; ++j ) {
-	mi = i;
-	if ( i==(cvt_rect)/2 && ((j==0 && rectelipse) || (j==1 && polystar)) )
-	    ++mi;
-/*	if ( cv->b.sc->parent->order2 && buttons[mi][j]==&GIcon_freehand ) */
-/*	    GDrawDrawImage(pixmap,&GIcon_greyfree,NULL,j*27+1,i*27+1);	 */
-/*	else								 */
-	    GDrawDrawImage(pixmap,buttons[mi][j],NULL,j*27+1,i*27+1);
-	norm = (mi*2+j!=tool);
-	{
-	  // These are from charview.c.
-	  extern int cvbutton3d; // Default 1.
-	  extern Color cvbutton3dedgelightcol; // Default 0xe0e0e0.
-	  extern Color cvbutton3dedgedarkcol; // Default 0x707070.
-	  if (cvbutton3d) {
-	    GDrawDrawLine(pixmap,j*27,i*27,j*27+25,i*27,norm?cvbutton3dedgelightcol:cvbutton3dedgedarkcol);
-	    GDrawDrawLine(pixmap,j*27,i*27,j*27,i*27+25,norm?cvbutton3dedgelightcol:cvbutton3dedgedarkcol);
-	    GDrawDrawLine(pixmap,j*27,i*27+25,j*27+25,i*27+25,norm?cvbutton3dedgedarkcol:cvbutton3dedgelightcol);
-	    GDrawDrawLine(pixmap,j*27+25,i*27,j*27+25,i*27+25,norm?cvbutton3dedgedarkcol:cvbutton3dedgelightcol);
```

Comment:
every button has two images, one for clicked, one for non clicked. If you want the relief then the images contain them. If you want to use the color to show selected tool and avoid the clutter of the boxes then you are free to do that too.